controllers/notificationpolicy_controller.go

```diff
@@ -135,12 +157,104 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
 		return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
 	}
 
+	if len(mergedRoutes) > 0 {
+		status := statusDiscoveredRoutes(mergedRoutes)
+		notificationPolicy.Status.DiscoveredRoutes = &status
+	}
+
+	if err := r.updateNotificationPolicyRoutesStatus(ctx, notificationPolicy, mergedRoutes); err != nil {
+		r.Log.Error(err, "failed to add merged events to routes")
+	}
+
+	if notificationPolicy.Spec.Route.IsRouteSelectorMutuallyExclusive() {
+		meta.RemoveStatusCondition(&notificationPolicy.Status.Conditions, conditionRoutesIgnoredDueToRouteSelector)
+	} else {
+		setRoutesIgnoredDueToRouteSelectorCondition(&notificationPolicy.Status.Conditions, notificationPolicy.Generation)
+	}
+
 	condition := buildSynchronizedCondition("Notification Policy", conditionNotificationPolicySynchronized, notificationPolicy.Generation, applyErrors, len(instances))
 	meta.SetStatusCondition(&notificationPolicy.Status.Conditions, condition)
 
 	return ctrl.Result{RequeueAfter: notificationPolicy.Spec.ResyncPeriod.Duration}, nil
 }
 
+// assembleNotificationPolicyRoutes iterates over all routeSelectors transitively.
+// returns an assembled GrafanaNotificationPolicy as well as a list of all merged routes.
+// it ensures that there are no reference loops when discovering routes via labelSelectors
+
+func assembleNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, namespace *string, notificationPolicy *grafanav1beta1.GrafanaNotificationPolicy) (*grafanav1beta1.GrafanaNotificationPolicy, []*v1beta1.GrafanaNotificationPolicyRoute, error) {
+	assembledPolicy := notificationPolicy.DeepCopy()
+	mergedRoutes := []*v1beta1.GrafanaNotificationPolicyRoute{}
+
+	// visitedGlobal keeps track of all routes that have been appended to mergedRoutes
+	// so we can record a status update for them later
+	visitedGlobal := make(map[string]bool)
+
+	// visitedChilds keeps track of all routes that have been visited on the current path
+	// so we can detect loops
+	visitedChilds := make(map[string]bool)
+
+	var assembleRoute func(*grafanav1beta1.Route) error
+	assembleRoute = func(route *grafanav1beta1.Route) error {
+		if route.RouteSelector != nil {
+			routes, err := getMatchingNotificationPolicyRoutes(ctx, k8sClient, route.RouteSelector, namespace)
+			if err != nil {
+				return fmt.Errorf("failed to get matching routes: %w", err)
+			}
+
+			// Replace the RouteSelector with matched routes
+			route.RouteSelector = nil
+			for i := range routes.Items {
+				matchedRoute := &routes.Items[i]
+				key := fmt.Sprintf("%s/%s", matchedRoute.Namespace, matchedRoute.Name)
```

Comment:
nitpick: I'd prefer to implement the `NamespacedResource` for routes and use it here